### ISSUE TITLE
Data Concatenation: how to avoid sample contamination during training

### ISSUE THREAD

**Issue by xiabingquan** [VITA-MLLM/VITA#25]
Hello VITA team. 

Thanks for this great work. After reading through the code and the preprint paper.  I got a question of data concatenation.

The paper mentions that you use "Data Concanetation" technique to concatenate different samples into a sequence. In this case, to avoid different samples attend to each other, the causal mask should be modified. 

For example, if a sequence with length 3 contains two samples with lengths 1 and 2. 
The original causal mask is:
```python
[
  [1, 0, 0],
  [1, 1, 0],
  [1, 1, 1],
]
```
And the modified mask should be:
```python
[
  [1, 0, 0],
  [0, 1, 0],
  [0, 1, 1],
]
```

I noticed that this codebase supports three attention implementations: eager、SDPA and Flash Attention.
However, customized attention mask is not supported in Flash Attention (see [issue](https://github.com/Dao-AILab/flash-attention/issues/352)).

My question is: 
What attention implementation do you use for training? Do you consider the sample contamination problem when using data concatenation? Thanks!

**Comment by xiabingquan** [comment]
I noticed that Flash Attention is used when training mixtral-8x7b
https://github.com/VITA-MLLM/VITA/blob/b546cd070c20142e400cff5179413f5bd70e90a9/vita/train/train.py#L255-L263
Is this setting also used in training smaller models?

> Currently, the model checkpoints are not immediately available to me. Any replies will help, thanks!

**Comment by linhaojia13** [comment]
Thanks for your attention. It is a good idea to modify the attention mask to avoid different samples attend to each other. However, we have not implemented such a mechanism because we found that the negative impact of simple concatenation is relatively small in our experiments. 

We still believe that modifying the attention mask is necessary. If you come up with a solution in Flash Attention,  welcome a pull request!

**Comment by xiabingquan** [comment]
> Thanks for your attention. It is a good idea to modify the attention mask to avoid different samples attend to each other. However, we have not implemented such a mechanism because we found that the negative impact of simple concatenation is relatively small in our experiments.
> 
> We still believe that modifying the attention mask is necessary. If you come up with a solution in Flash Attention, welcome a pull request!

Got it. Thanks for your quick reply. A workaround is using customized flash attention kernels based on Trion, such as [flashattention2-custom-mask](https://github.com/alexzhang13/flashattention2-custom-mask), but its correctntess is not tested.

**Comment by xiabingquan** [comment]
@linhaojia13 The function [`flash_attn_varlen_func`](https://github.com/Dao-AILab/flash-attention/blob/8c20cfef497aa4e88940b92476a6a625e9fc87f5/flash_attn/flash_attn_interface.py#L1051-L1141) in [flash attention](https://github.com/Dao-AILab/flash-attention/tree/main/flash_attn) could meet our needs (with some modifcations to cu_lens). 

I've verified its correctness. Hope it helps! 😀

**Comment by linhaojia13** [comment]
Thank you!
